app/src/lib/components/Select.svelte

```diff
@@ -1,18 +1,30 @@
-<script lang="ts">
+<script lang="ts" context="module">
+	export type SelectItemType<S extends string> = Record<S, unknown>;
+</script>
+
+<script lang="ts" generics="SelectItemType extends Record<string, unknown>">
 	import ScrollableContainer from './ScrollableContainer.svelte';
 	import TextBox from './TextBox.svelte';
 	import { clickOutside } from '$lib/clickOutside';
+	import { filterStringByKey } from '$lib/utils/filters';
+	import { KeyName } from '$lib/utils/hotkeys';
+	import { throttle } from '$lib/utils/misc';
 	import { pxToRem } from '$lib/utils/pxToRem';
+	import { isChar } from '$lib/utils/string';
 	import { createEventDispatcher } from 'svelte';
 
+	const INPUT_THROTTLE_TIME = 100;
+
+	type SelectItemKeyType = keyof SelectItemType;
```

Comment:
> Since it's already a type, could we drop the `Type` suffix?
>
Totally 👍 

> But also, since it's a keyof something that extends Record<string, unknown>, I don't see the value of this type.
>
It's so that we can narrow down the `labelId` and `itemId` properties (see lines 26 and 27). This way, the component dynamically checks what the item type is, and based on that we set the acceptable types for the `labelId` and `itemId`.

What this does in practice is for the `AIPromptSelect.svelte` component is that if we pass the `allPrompts: UserPrompt[]` variable as the `items` property, the component will infer that the only allowed keys are `"id"`, `"name"` or `"prompt"`

```ts
export type UserPrompt = {
	id: string;
	name: string;
	prompt: Prompt;
};
```

Which results in the auto-completion in the IDE:
<img width="443" alt="auto-suggest" src="https://github.com/gitbutlerapp/gitbutler/assets/35891811/1a52737e-23f1-4db9-baeb-38d3fcdad42e">

And if the wrong key is used or if it's updated, a TS compiler error will catch that:
<img width="568" alt="type-guard" src="https://github.com/gitbutlerapp/gitbutler/assets/35891811/529bf50e-e33a-415c-8820-927b5a755431">

Reply:
Thanks, you're right. That's awesome.